cpp/include/cudf/datetime.hpp

```diff
@@ -182,7 +182,7 @@ std::unique_ptr<cudf::column> day_of_year(
  *
  * @returns cudf::column of timestamp type containing the computed timestamps.
  * @throw cudf::logic_error if `timestamps` datatype is not a TIMESTAMP or if `months` datatype
- * is not INT16.
+ * is not INT16 or INT32.
```

Comment:
can't we just allow any integer type?

Reply:
Sure. Especially now using indexalator we have practically already supported it.

Reply:
But wait, I spoke too soon.
There is a potential issue in that the type gets implicitly converted into `int32_t` at line 256:
```c++
ymd += duration_m{months_val};
```
where `duration_m` is
```c++
using duration_m = duration<int32_t, months::period>;
```
We should use int64_t here if we are to also support integer type larger than int32_t. But for unsigned types, do we want to implicitly cast them into signed type too?

Moreover, do we even want to support unsigned types here? Negative months can be used as subtracting months from the given date, when user passes in an unsigned column it limits the function to only performing addition. Not a big deal of course, but just wonder if it allows potential user error. Maybe @revans2 has better insight.

Reply:
For me I have a set of specific use cases that I am trying to support.

1. The Spark `AddMonths` operator
2. The Spark `CalendarIntervalType`
3. The upcomming Spark `YearMonthIntervalType`

In all of these cases the number of months to add is represented as an int32_t value.  Anything beyond that is not something that I need right now. I think the reason for this is because calculating an exact date or timestamp that is more than 178 million years in the future or the past feels like a stretch for any use case. Despite the fact that TIMESTAMP_SECONDS technically could handle more than that without over or under flowing.

Reply:
This reminds me that if we are to support `int64_t` type, then `timestamp_s` type may overflow with a very large months value.

Reply:
The reason we want to support any type of int is for generality---we may have timestamps stored in some int type column (such as int64) and they need to be cast into int32 before input into this API. We can avoid such casting and use that column directly---the caller is responsible for checking under/overflow.

Reply:
I understand the reason for generality, but there is also the question of combinatorics. I am not an expert on how this code is being written, but we it can be done in such a way that we don't end up with 40 kernels compiled for just this one simple op, then generality is fine with me.

Reply:
cudf has `indexalator` that can normalize the input int type into int32. This does involve type dispatching in the host code with a number of code paths equal to the number of int types. There will be just one kernel compiled for int32.

Reply:
There is also one thing I'm concern about: most existing date/time APIs already dealt with just int16 or int32. We should enforce consistency by either:
 * Just support int16 and int32, as the existing APIs do, or
 *  Add general int types support in this PR, and file an issue requesting to add general int types support for all other date/time APIs

Reply:
> cudf has indexalator that can normalize the input int type into int32. This does involve type dispatching in the host code with a number of code paths equal to the number of int types. There will be just one kernel compiled for int32.

Thanks, I almost forget about this. Maybe we shouldn't support column beyond `int32_t` then? Because there's the potential implicit truncation.

> There is also one more thing I'm concerned about: most existing date/time APIs already dealt with just int16 or int32.

So far we only have one binary datetime processing method (this). Other functions are unary ops that accepts any timestamp typed columns. The left operands of this method matches with the type of the lone argument to the rest methods, and it's the right operand we are discussing here. Because we don't have anywhere else to look at, we just deduce to the best of our knowledge here. The `int32_t` intuition tells me that maybe we shouldn't lie that we support `int64_t` but actually we are performing a truncation under the hood. Nor we should support unsigned types because we are doing an implicit cast. We could support `int8_t` here if that's necessary.

> we don't end up with 40 kernels compiled for just this one simple op

There are two type dispatching happening in here - one for months argument and one for the timestamp column. And (thankfully to the indexalator magic) they are uncorrelated so we don't have combinatory problems.